### ISSUE TITLE
ipv6: rpc socket bind uses wrong source address

### ISSUE THREAD

**Issue by zimnx** [scylladb/scylla#5175]
*Installation details*
Scylla version (or git commit hash): 88611d41d08
Cluster size: 6 nodes, 2 DCs (3 nodes per dc)
OS (RHEL/CentOS/Ubuntu/AWS AMI): docker on ubuntu host

I'm trying to setup IPv6 cluster inside docker. I configured `fd00::/64` subnet with following cluster setup:
```
dc1:
node1: fd00::1:1
node2: fd00::1:2
node3: fd00::1:3

dc2:
node1: fd00::2:1
node2: fd00::2:2
node3: fd00::2:3
```

I added following entries to scylla.yaml config file:
```
enable_ipv6_dns_lookup: true
listen_interface: eth0
listen_interface_prefer_ipv6: true
rpc_interface: eth0
rpc_interface_prefer_ipv6: true
```

Every container has ipv4 and ipv6 address on `eth0` interface.
I'm starting scylla with following arguments:
```
# from dc1 node3
/usr/bin/scylla --log-to-syslog 0 --log-to-stdout 1 --default-log-level info --network-stack posix --developer-mode=1 --smp 2 --overprovisioned --listen-address fd00::1:3 --rpc-address fd00::1:3 --seed-provider-parameters seeds=fd00::1:1,fd00::2:1 --blocked-reactor-notify-ms 999999999
```

Nodes are not able to form cluster. On every node logs I see these entries:
```
INFO  2019-10-15 14:54:33,411 [shard 0] gossip - Connect seeds again ... (182 seconds passed)
INFO  2019-10-15 14:54:33,411 [shard 0] rpc - client [fd00::1:1]:7000: fail to connect: bind: Address family not supported by protocol
INFO  2019-10-15 14:54:33,412 [shard 0] rpc - client [fd00::2:1]:7000: fail to connect: bind: Address family not supported by protocol
```
and after multiple failures scylla process restarts.

Probably there's a bug inside rpc/seastar because socket is trying to bind to wrong address. `bind` `sin_addr` is using wrong address family.

`strace` snippet below:
```
socket(AF_INET6, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_TCP) = 30                                                                  
setsockopt(30, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0                                                                                        
setsockopt(30, SOL_IP, 0x18 /* IP_??? */, [1], 4) = 0                                                                                       
bind(30, {sa_family=AF_INET, sin_port=htons(51024), sin_addr=inet_addr("0.0.0.0")}, 128) = -1 EAFNOSUPPORT (Address family not supported by protocol)
```

**Comment by slivne** [comment]
scylla doesn't support listen_interface / rpc_interface

https://github.com/scylladb/scylla/issues/1058

please use listen_address and rpc_address and specify the ipv6 value

**Comment by zimnx** [comment]
@slivne I'm passing `listen_address` and `rpc_address` as a command line argument. Without `listen_interface` and `rpc_interface` I'm seeing same issue.

**Comment by slivne** [comment]
please remove the interface configs

listen_interface: eth0
listen_interface_prefer_ipv6: true
rpc_interface: eth0
rpc_interface_prefer_ipv6: true


does this work for you ?


if not ... @Calle Wilund <calle@scylladb.com> any idea ?

**Comment by slivne** [comment]
@zimnx I need the scylla.yaml and the ``` ps -ef | grep scylla ``` that you ran to try and figure this out

I was able to run scylla with ipv6 https://github.com/scylladb/scylla/issues/5177#issuecomment-542438144

**Comment by zimnx** [comment]
@slivne 
my scylla.yaml: https://gist.github.com/zimnx/59538ec6e6987728aa9ac9d9d3bb58c2
ps output:
```
[root@995bdcbcbb55 /]# ps -ef | grep scylla
root         1     0  0 07:50 ?        00:00:00 /opt/scylladb/python3/bin/python3 /opt/scylladb/python3/bin/../libexec/python3.7.bin -s /docker-entrypoint.py --smp 2 --seeds fd00::1:1,fd00::2:1 --listen-address fd00::1:3 --rpc-address fd00::1:3                                                                                                                                                                            
root        27    24  1 07:50 ?        00:00:00 /usr/bin/scylla --log-to-syslog 0 --log-to-stdout 1 --default-log-level info --network-stack posix --developer-mode=1 --smp 2 --overprovisioned --listen-address fd00::1:3 --rpc-address fd00::1:3 --seed-provider-parameters seeds=fd00::1:1,fd00::2:1 --blocked-reactor-notify-ms 999999999                                                                                   
root        28    24  0 07:50 ?        00:00:00 /bin/bash /scylla-housekeeping-service.sh
root        30    24  1 07:50 ?        00:00:01 /opt/scylladb/scripts/jmx/symlinks/scylla-jmx -Xmx256m -XX:+UseSerialGC -XX:+HeapDumpOnOutOfMemoryError -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.host=localhost -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=7199 -Djava.rmi.server.hostname=localhost -Dcom.sun.management.jmxremote.rmi.port=7199 -Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder -jar /opt/scylladb/scripts/jmx/scylla-jmx-1.0.jar
root       177    83  0 07:51 pts/0    00:00:00 grep --color=auto scylla
```

Additionally ifconfig from one of the nodes:
```
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.19.0.4  netmask 255.255.0.0  broadcast 172.19.255.255
        inet6 fd00::1:3  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::42:acff:fe13:4  prefixlen 64  scopeid 0x20<link>
        ether 02:42:ac:13:00:04  txqueuelen 0  (Ethernet)
        RX packets 216  bytes 33126 (32.3 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 70  bytes 11328 (11.0 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 94  bytes 9473 (9.2 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 94  bytes 9473 (9.2 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

```

**Comment by slivne** [comment]
I expect the issue is that rpc/seastar does not check the socket family and blindly binds to the ipv4 "0.0.0.0" 

#3066 added support for nodes that support multi-nics and relies on the OS to route correctly - as such we do not pass the listen_address any more (which is set to ipv6) so the listen_address does not influence what is used.

**Comment by elcallio** [comment]
rpc connect issue is (at least partially) due to: scylladb/seastar#694

**Comment by elcallio** [comment]
patch sets for seastar fixing both scope and connect issues have been sent.

**Comment by zimnx** [comment]
@elcallio where can I track it?

**Comment by slivne** [comment]
seastar patch merged to scylla 771e028c1aafb86a870fbffd845cbb8fa2f93564
